### ISSUE TITLE
Restarting a Causal Cluster member with a different IP

### ISSUE THREAD

**Issue by thessem** [neo4j/neo4j#9326]
I have a 3 member Causal Cluster setup within Kubernetes with the following configuration
```
dbms.mode=CORE\n
dbms.connectors.default_advertised_address=$HOSTNAME.neo4j.infrastructure.svc.cluster.local\n
causal_clustering.expected_core_cluster_size=3
causal_clustering.initial_discovery_members=neo4j-core-0.neo4j.infrastructure.svc.cluster.local:5000,neo4j-core-1.neo4j.infrastructure.svc.cluster.local:5000,neo4j-core-2.neo4j.infrastructure.svc.cluster.local:5000
causal_clustering.discovery_advertised_address=$HOSTNAME.neo4j.infrastructure.svc.cluster.local:5000
causal_clustering.transaction_advertised_address$HOSTNAME.neo4j.infrastructure.svc.cluster.local:6000
causal_clustering.raft_advertised_address=$HOSTNAME.neo4j.infrastructure.svc.cluster.local:7000
```
where `$HOSTAME` is `neo4j-core-0`, `neo4j-core-1` or `neo4j-core-2`.

Currently, if I simulate an outage by stopping one of the pods running on Kubernetes it will restart the pod with the same `$HOSTNAME` but a different IP. In this case it is `neo4j-core-1` restarting. The logs for the new member look like this

```
2017-05-10 06:56:35.099+0000 INFO  ======== Neo4j 3.1.3 ========
2017-05-10 06:56:35.742+0000 INFO  Starting...
2017-05-10 06:56:37.297+0000 INFO  Bolt enabled on 0.0.0.0:7687.
2017-05-10 06:56:37.329+0000 INFO  Initiating metrics...
2017-05-10 06:56:37.657+0000 INFO  My connection info: [
	Discovery:   listen=0.0.0.0:5000, advertised=neo4j-core-1.neo4j.infrastructure.svc.cluster.local:5000,
	Transaction: listen=0.0.0.0:6000, advertised=neo4j-core-1.neo4j.infrastructure.svc.cluster.local:6000, 
	Raft:        listen=0.0.0.0:7000, advertised=neo4j-core-1.neo4j.infrastructure.svc.cluster.local:7000, 
	Client Connector Addresses: bolt://neo4j-core-1.neo4j.infrastructure.svc.cluster.local:7687,http://neo4j-core-1.neo4j.infrastructure.svc.cluster.local:7474,https://neo4j-core-1.neo4j.infrastructure.svc.cluster.local:7473
]
2017-05-10 06:56:37.662+0000 INFO  Discovering cluster with initial members: [neo4j-core-0.neo4j.infrastructure.svc.cluster.local:5000, neo4j-core-1.neo4j.infrastructure.svc.cluster.local:5000, neo4j-core-2.neo4j.infrastructure.svc.cluster.local:5000]
2017-05-10 06:56:37.662+0000 INFO  Attempting to connect to the other cluster members before continuing...
2017-05-10 07:07:04.218+0000 ERROR Failed to start Neo4j: Starting Neo4j failed: Component 'org.neo4j.server.database.LifecycleManagingDatabase@4c57196a' was successfully initialized, but failed to start. Please see attached cause exception. Starting Neo4j failed: Component 'org.neo4j.server.database.LifecycleManagingDatabase@4c57196a' was successfully initialized, but failed to start. Please see attached cause exception.
org.neo4j.server.ServerStartupException: Starting Neo4j failed: Component 'org.neo4j.server.database.LifecycleManagingDatabase@4c57196a' was successfully initialized, but failed to start. Please see attached cause exception.
	at org.neo4j.server.exception.ServerStartupErrors.translateToServerStartupError(ServerStartupErrors.java:68)
	at org.neo4j.server.AbstractNeoServer.start(AbstractNeoServer.java:230)
	at org.neo4j.server.ServerBootstrapper.start(ServerBootstrapper.java:91)
	at org.neo4j.server.ServerBootstrapper.start(ServerBootstrapper.java:68)
	at org.neo4j.server.enterprise.EnterpriseEntryPoint.main(EnterpriseEntryPoint.java:32)
Caused by: org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.server.database.LifecycleManagingDatabase@4c57196a' was successfully initialized, but failed to start. Please see attached cause exception.
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:443)
	at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
	at org.neo4j.server.AbstractNeoServer.start(AbstractNeoServer.java:202)
	... 3 more
Caused by: java.lang.RuntimeException: Error starting org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory, /var/lib/neo4j/data/databases/graph.db
	at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.initFacade(GraphDatabaseFacadeFactory.java:199)
	at org.neo4j.causalclustering.core.CoreGraphDatabase.<init>(CoreGraphDatabase.java:56)
	at org.neo4j.causalclustering.core.CoreGraphDatabase.<init>(CoreGraphDatabase.java:47)
	at org.neo4j.server.enterprise.EnterpriseNeoServer.lambda$static$2(EnterpriseNeoServer.java:96)
	at org.neo4j.server.database.LifecycleManagingDatabase.start(LifecycleManagingDatabase.java:89)
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:433)
	... 5 more
Caused by: org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.causalclustering.core.consensus.membership.MembershipWaiterLifecycle@27ab572b' was successfully initialized, but failed to start. Please see attached cause exception.
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:443)
	at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
	at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.initFacade(GraphDatabaseFacadeFactory.java:195)
	... 10 more
Caused by: java.lang.RuntimeException: Server failed to join cluster within catchup time limit [600000 ms]
	at org.neo4j.causalclustering.core.consensus.membership.MembershipWaiterLifecycle.start(MembershipWaiterLifecycle.java:69)
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:433)
	... 12 more
Caused by: java.util.concurrent.TimeoutException
	at java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1771)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
	at org.neo4j.causalclustering.core.consensus.membership.MembershipWaiterLifecycle.start(MembershipWaiterLifecycle.java:57)
	... 13 more
```
Where I'm assuming the crash is caused by a 10m timeout. The server will never rejoin the cluster.

Looking at the debug logs for one of the servers still running in the cluster gives me
```
2017-05-10 06:56:34.799+0000 INFO  [o.n.c.m.RaftOutbound] No address found for MemberId{05c30913}, probably because the member has been shut down.
2017-05-10 06:56:37.485+0000 INFO  [o.n.c.c.c.m.RaftMembershipManager] Target membership: [MemberId{7dc11ef3}, MemberId{6e6d2545}]
2017-05-10 06:56:37.489+0000 INFO  [o.n.c.d.HazelcastCoreTopologyService] Current read replica topology is ReadReplicaTopology{readReplicas=[]}
2017-05-10 06:56:41.774+0000 INFO  [o.n.c.d.HazelcastCoreTopologyService] Core member added MembershipEvent {member=Member [neo4j-core-1.neo4j.infrastructure.svc.cluster.local]:5000 - f1a361a7-e31a-4b11-81fd-b4b4c84a9bb6,type=added}
2017-05-10 06:56:41.778+0000 INFO  [o.n.c.c.c.m.RaftMembershipManager] Target membership: [MemberId{7dc11ef3}, MemberId{05c30913}, MemberId{6e6d2545}]
2017-05-10 06:56:41.857+0000 INFO  [o.n.c.d.HazelcastCoreTopologyService] Current read replica topology is ReadReplicaTopology{readReplicas=[]}
```

where `05c30913` is the shutdown member, and `f1a361a7` is `neo4j-core-1` and is the new member of the cluster. Also in this log are constant elections
```
2017-05-10 07:03:28.391+0000 INFO  [o.n.c.c.c.RaftMachine] Election timeout triggered
2017-05-10 07:03:28.391+0000 INFO  [o.n.c.c.c.RaftMachine] Election started with vote request: Vote.Request from MemberId{7dc11ef3} {term=229, candidate=MemberId{7dc11ef3}, lastAppended=87, lastLogTerm=227} and members: [MemberId{05c30913}, MemberId{7dc11ef3}, MemberId{6e6d2545}]
2017-05-10 07:03:28.391+0000 INFO  [o.n.c.c.c.RaftMachine] Moving to CANDIDATE state after successfully starting election
2017-05-10 07:03:28.405+0000 INFO  [o.n.c.c.c.RaftMachine] Moving to LEADER state at term 229 (I am MemberId{7dc11ef3}), voted for by [MemberId{6e6d2545}]
2017-05-10 07:03:28.406+0000 INFO  [o.n.c.c.c.s.RaftState] Leader changed from MemberId{6e6d2545} to MemberId{7dc11ef3}
2017-05-10 07:03:28.407+0000 INFO  [o.n.c.c.c.s.RaftLogShipper] Starting log shipper: MemberId{05c30913}[matchIndex: -1, lastSentIndex: 0, localAppendIndex: 88, mode: MISMATCH]
2017-05-10 07:03:28.408+0000 INFO  [o.n.c.c.c.s.RaftLogShipper] Starting log shipper: MemberId{6e6d2545}[matchIndex: -1, lastSentIndex: 0, localAppendIndex: 88, mode: MISMATCH]
2017-05-10 07:03:28.408+0000 INFO  [o.n.c.c.c.m.RaftMembershipChanger] Idle{}
2017-05-10 07:03:28.413+0000 INFO  [o.n.c.c.c.s.RaftLogShipper] MemberId{6e6d2545}[matchIndex: 87, lastSentIndex: 88, localAppendIndex: 88, mode: MISMATCH]: caught up after mismatch, moving to PIPELINE mode
2017-05-10 07:03:32.485+0000 INFO  [o.n.c.c.c.m.RaftMembershipManager] Target membership: [MemberId{7dc11ef3}, MemberId{05c30913}, MemberId{6e6d2545}]
2017-05-10 07:03:32.489+0000 INFO  [o.n.c.d.HazelcastCoreTopologyService] Current read replica topology is ReadReplicaTopology{readReplicas=[]}
2017-05-10 07:03:37.485+0000 INFO  [o.n.c.c.c.m.RaftMembershipManager] Target membership: [MemberId{7dc11ef3}, MemberId{05c30913}, MemberId{6e6d2545}]
2017-05-10 07:03:37.489+0000 INFO  [o.n.c.d.HazelcastCoreTopologyService] Current read replica topology is ReadReplicaTopology{readReplicas=[]}
2017-05-10 07:03:38.962+0000 INFO  [o.n.c.c.c.RaftMachine] Moving to FOLLOWER state after receiving vote request at term 230 (my term is 229) from MemberId{05c30913}
2017-05-10 07:03:38.965+0000 INFO  [o.n.c.c.c.s.RaftState] Leader changed from MemberId{7dc11ef3} to null
2017-05-10 07:03:38.965+0000 INFO  [o.n.c.c.c.s.RaftLogShipper] Stopping log shipper MemberId{05c30913}[matchIndex: -1, lastSentIndex: 88, localAppendIndex: 88, mode: MISMATCH]
2017-05-10 07:03:38.965+0000 INFO  [o.n.c.c.c.s.RaftLogShipper] Stopping log shipper MemberId{6e6d2545}[matchIndex: 88, lastSentIndex: 88, localAppendIndex: 88, mode: PIPELINE]
2017-05-10 07:03:38.966+0000 INFO  [o.n.c.c.c.m.RaftMembershipChanger] Inactive{}
2017-05-10 07:03:41.411+0000 INFO  [o.n.c.c.c.RaftMachine] Election timeout triggered
```
but running `CALL dbms.cluster.overview()` on one of the older members still references the old ID of `neo4j-core-1` which was `05c30913`.

I'd like to know if this is really because the server is changing IP address and that is not supported (if so consider this a feature request) or if it's because of a configuration issue I have.

**Comment by dannolan** [comment]
👍 any information on this?

**Comment by dannolan** [comment]
This is apparently fixed in Neo4J version 3.1.6 according to @mneedham 

@thessem can you close this?

**Comment by thessem** [comment]
Can confirm this is fixed in 3.2.3
